README.md
# intakeinformer

A nutritional tracker wrapped around [the FoodData Central USDA API](https://fdc.nal.usda.gov/api-guide.html). The wrapper allows users to track their daily food intake and analyze the nutritional content of their diet based on the USDA food database and their own selections.

#### The two core functionalities of intakeinformer – get_calories_for_food_query() and dri_benchmark() – produce distinct outputs:
- **Calorie Calculation & Comparison Visualization**: '**get_calories_for_food_query()**' calculates and returns the calorie content of a user-selected food query, normalized to a standard serving size of 100g. The calorie calculation considers the three main macronutrients (carbohydrates, fats, and proteins) of the food and generates a visualization (bar chart) depicting the calorie content per 100g of selected food items alongside a cumulative total. This visual aid helps users intuitively understand the calorie distribution of their food choices.
- **Nutrient Intake vs. DRI Recommendations**: '**dri_benchmark()**' produces a comparative bar chart that contrasts the actual nutrient intake from selected food items against [Dietary Reference Intakes (DRI)](https://www.ncbi.nlm.nih.gov/books/NBK545442/table/appJ_tab3/?report=objectonly) recommendations for a user-specified life stage. This output is instrumental in evaluating dietary choices against nutritional benchmarks.

## Installation

```python
     pip install intakeinformer
```
## Example Usage
- Because the interactive process for generating outputs with the two core functionalities of intakeinformer – **get_calories_for_food_query()** and **dri_benchmark()**. Example code & output snippets are insufficient to show the complete usage of the package's main functionalities.

Please check out an [Example Usage Showcase](docs/example.ipynb) here to see two of IntakeInformer's core functionalities in action!

## Getting Started with IntakeInformer

### Step 1: Obtain a USDA FoodData Central API Key
1. **Register for an API Key:**
   - Visit the USDA FoodData Central API's [registration page](https://fdc.nal.usda.gov/api-key-signup.html).
   - Fill in the required details and submit the form to register.
   - Once registered, you will receive an API key. Keep this key secure as it is your personal access token to the API.

### Step 2: Set the API Key as an Environment Variable
Setting the API key as an environment variable keeps it secure and easily accessible by the package.

**For Unix/Linux/macOS:**
1. Open your terminal.
2. Run the following command, replacing `'your_api_key_here'` with the actual API key you received:
   ```bash
   export POETRY_FDC_API_KEY='your_api_key_here'
3. To ensure the variable is set, you can run
   ```bash
   echo $POETRY_FDC_API_KEY
   ```
   It should display your API key.

**For Windows:**

1. **Open Command Prompt:**
   - You can search for "cmd" or "Command Prompt" in the Windows search bar and open it.

2. **Set the Environment Variable:**
   - Run the following command in Command Prompt:
     ```cmd
     set POETRY_FDC_API_KEY=your_api_key_here
     ```
   - Replace `your_api_key_here` with your actual API key.

3. **Verify the Environment Variable:**
   - To check if the environment variable is set correctly, run:
     ```cmd
     echo %POETRY_FDC_API_KEY%
     ```
   - This should display your API key.

### Step 3: Install and Use IntakeInformer
1. **Install the Package:**
   - Use pip (or poetry if you're using it) to install `IntakeInformer`.
   - Command: `pip install intakeinformer` (or the appropriate command if using poetry).

2. **Using IntakeInformer:**
   - Import and use the functions from IntakeInformer in your Python scripts or interactive environment.
   - Example:
     ```python
     from intakeinformer import get_calories_for_food_query, dri_benchmark

     # Use the functions as needed
     ```

### Step 4: Troubleshooting
- **API Key Issues:** If you encounter issues related to the API key (such as it not being recognized), ensure that you have correctly set the environment variable and that your current session has access to it.
- **Environment Variables:** Remember that environment variables set in a terminal session are usually temporary. If you restart your system, you might need to set the environment variable again, unless you add it to your system's permanent profile.

### Step 5: Seeking Help
- If you run into any issues or have questions, refer to the documentation of `IntakeInformer` or reach out for support as provided by the package.


## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`intakeinformer` was created by Junchen Xiong. It is licensed under the terms of the MIT license.

## Credits

`intakeinformer` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (11/12/2023)

- First release of `intakeinformer`!

- ## v0.1.1 (13/12/2023)

- Minor tool.poetry.dependencies updates.

- ## v0.1.2 (14/12/2023)
- Minor function update in data_fetching.py file.

-  ## v0.1.3 (14/12/2023)
- Minor function change in data_fetching.py file.
- Test FDC_Key input from users.

-  ## v0.1.4 (14/12/2023)
- Modify the _fetch_food_data function to accept an additional parameter for the API key.
- Modify your main functions (get_calories_for_food_query and dri_benchmark) to include the API key as a parameter and pass it to _fetch_food_data.

-  ## v0.1.5 (14/12/2023)
- Minor Fix to the previous version.

-  ## v0.1.6 (14/12/2023)
- Minor Fix.

- ## v1.0.0 (14/12/2023)
- Read the Docs Remote Rendering Ready.
- Set up testing files for selected functions.
- PyPi ready.

tests/test_intakeinformer.py
from intakeinformer import intakeinformer


tests/test_main_functions.py
from intakeinformer import intakeinformer

'''
test the main functionalities provided by get_calories_for_food_query() and dri_benchmark() functions. Testing these functions can be a bit challenging because they involve user input and API calls. 
'''

import unittest
from unittest.mock import patch
from intakeinformer.main_functions import get_calories_for_food_query, dri_benchmark
import pandas as pd

class TestMainFunctions(unittest.TestCase):

    @patch('intakeinformer.main_functions._fetch_food_data')
    @patch('intakeinformer.main_functions.get_user_food_selection')
    def test_get_calories_for_food_query(self, mock_selection, mock_fetch):
        # Mock the _fetch_food_data function to return a predefined DataFrame
        mock_fetch.return_value = pd.DataFrame({'description': ['Apple', 'Banana'], 
                                                'foodNutrients': [{'foodNutrientId': 1008, 'value': 52},
                                                                  {'foodNutrientId': 1008, 'value': 89}]})

        # Mock the get_user_food_selection function to return a predefined selection
        mock_selection.side_effect = lambda x: x.iloc[0]

        # Call the function
        result = get_calories_for_food_query("apple", "banana")

        # Define the expected result
        expected_result = {'Apple': 52.0, 'Banana': 89.0, 'total': 141.0}

        # Check if the result matches the expected result
        self.assertEqual(result, expected_result)

    @patch('intakeinformer.main_functions._fetch_food_data')
    @patch('intakeinformer.main_functions.get_user_food_selection')
    @patch('intakeinformer.main_functions.get_user_life_stage')
    def test_dri_benchmark(self, mock_life_stage, mock_selection, mock_fetch):
        # Similar to the above, mock the dependent functions
        # And then call dri_benchmark with the mock data
        # Check if the function performs as expected
        pass  # Implementation will depend on how dri_benchmark is structured

if __name__ == '__main__':
    unittest.main()


tests/test_extract_nutrients.py
import pytest
from intakeinformer.data_processing import extract_nutrients

def test_extract_nutrients():
    # Mock input for food_nutrients
    mock_nutrients = [
        {'nutrientId': 1004, 'nutrientName': 'Total lipid (fat)', 'value': 0.2},
        {'nutrientId': 1003, 'nutrientName': 'Protein', 'value': 1.1},
        {'nutrientId': 1005, 'nutrientName': 'Carbohydrate, by difference', 'value': 14.3},
        {'nutrientId': 1079, 'nutrientName': 'Fiber, total dietary', 'value': 2.4},
        {'nutrientId': 1087, 'nutrientName': 'Calcium, Ca', 'value': 5},
        {'nutrientId': 1089, 'nutrientName': 'Iron, Fe', 'value': 0.36},
        {'nutrientId': 1092, 'nutrientName': 'Potassium, K', 'value': 107},
        {'nutrientId': 1093, 'nutrientName': 'Sodium, Na', 'value': 1},
        {'nutrientId': 1104, 'nutrientName': 'Vitamin A, IU', 'value': 54},
        {'nutrientId': 1162, 'nutrientName': 'Vitamin C, total ascorbic acid', 'value': 6.2},
        {'nutrientId': 1114, 'nutrientName': 'Vitamin D', 'value': 0},
        {'nutrientId': 1175, 'nutrientName': 'Vitamin B-12', 'value': 0.00}
    ]

    # Expected output
    expected_output = {
        'Carbohydrate(g/d)': 14.3,
        'Total Fiber(g/d)': 2.4,
        'Fat(g/d)': 0.2,
        'Proteinb(g/d)': 1.1,
        'Calcium (mg/d)': 5,
        'Iron (mg/d)': 0.36,
        'Potassium (mg/d)': 107,
        'Sodium (mg/d)': 1,
        'Vitamin A(μg/d)a': 54,
        'Vitamin C(mg/d)': 6.2,
        'Vitamin D(μg/d)b,c': 0,
        'Vitamin B12(μg/d)': 0.00
    }

    # Run the function with the mock data
    output = extract_nutrients(mock_nutrients)

    # Check if the output matches the expected output
    assert output == expected_output, f"Expected {expected_output}, but got {output}"


src/intakeinformer/data_fetching.py
import requests
from requests.exceptions import HTTPError
import json
import os
import pandas as pd

from bs4 import BeautifulSoup


'''
internal helper function to fetch data from API and parse it into dataframe for future operation
'''
def _fetch_food_data(FDC_Key, query):
    """
    Fetches food data from the USDA FoodData Central API for a specified query.

    This function sends a GET request to the USDA FoodData Central API with a given food query and returns the results as a pandas DataFrame. It handles HTTP errors and returns an empty DataFrame if no data is found.

    Parameters
    ----------
    query : str
        The food query string to search for in the USDA FoodData Central API.

    Returns
    -------
    pandas.DataFrame
        A DataFrame containing the fetched food data. Each row represents a different food item, with columns for various food attributes.

    Raises
    ------
    HTTPError
        If the HTTP request to the USDA API fails.

    Exception
        For general exceptions that may occur during the API request.

    Examples
    --------
    >>> _fetch_food_data("apple")
    DataFrame containing data for various apple products, with columns such as 'description', 'brandOwner', and nutritional information.

    Notes
    -----
    The function uses an internal API key defined as FDC_key. Ensure this key is properly set before using the function. The function is intended for internal use within the package and is not exposed to the end users.

    """
    url_search = "https://api.nal.usda.gov/fdc/v1/foods/search"
    params_search = {
        "api_key": FDC_Key,
        
        "pageSize": 10,     # number of results to return, default to 1 for ease of access
        "pageNumber": 2,   # Page number to retrieve, default to 1
        "dataType": [ "Branded"],
        "sortBy": "publishedDate",
        "sortOrder": "asc",
        "query": query
    }
    data = {}  # Initialize data as an empty dictionary

    try:
        response_food_search = requests.get(url_search, params=params_search)
        response_food_search.raise_for_status()
        data = response_food_search.json()  # Update data with the response
    except HTTPError as http_err:
        print(f'HTTP error occurred: {http_err}')
    except Exception as err:
        print(f'Other error occurred: {err}')


    if 'foods' in data:
        '''
        pd.json_normalize is used to flatten the nested JSON structure into a DataFrame. 
        '''
        return pd.json_normalize(data, record_path=['foods'])
    else:
        return pd.DataFrame()




def get_user_food_selection(food_df):
    """
    Prompts the user to select a specific food item from a list presented based on the queried food.

    This function displays a list of food items retrieved from the USDA FoodData Central API, 
    allowing the user to select one. The list includes various brands or types of the food item queried. 
    The user's selection is returned for further processing, such as nutrient analysis or calorie calculation.

    Parameters
    ----------
    food_df : pandas.DataFrame
        A DataFrame containing a list of food items, obtained from the 
        USDA FoodData Central API. Each row represents a different variety or brand 
        of the queried food.

    Returns
    -------
    pandas.DataFrame
        A single-row DataFrame representing the user's selected food item.

    Examples
    --------
    >>> food_df = _fetch_food_data("apple")
    >>> selected_apple = get_user_food_selection(food_df)
    >>> print(selected_apple)

    Notes
    -----
    - The function works as an internal helper function for both main functions.
    """
    if food_df.empty:
        print("No foods found.")
        return None

    # Display the top results to the user
    for i, row in food_df.iterrows():
        brand_owner = row.get('brandOwner', 'No brand information')
        print(f"{i + 1}: {row['description']} ({brand_owner}, {row.get('servingSize', 'N/A')}g)")

    # Ask the user to make a selection
    try:
        choice = int(input("Enter the number of your choice: ")) - 1
    except ValueError:
        print("Invalid input. Please enter a number.")
        return None

    # Return the selected food item
    if 0 <= choice < len(food_df):
        return food_df.iloc[choice]
    else:
        print("Invalid selection.")
        return None

def get_user_life_stage():
    """
    Interactively prompts users to identify their life stage category and age range 
    for Dietary Reference Intake (DRI) comparison. 

    This function asks users to choose from predefined life stage categories and age ranges 
    that align with DRI standards. These categories include options for Infants, Children, 
    Males, Females, Pregnancy, and Lactation, with specific age ranges under each. 
    The user's selections are returned as a tuple, which can be used to retrieve appropriate 
    DRI benchmarks for nutrient comparison.

    Returns
    -------
    tuple
        A tuple containing two strings: the life stage category (e.g., 'Children') and 
        the specific age range within that category (e.g., '1–3 y').

    Examples
    --------
    >>> life_stage_info = get_user_life_stage()
    >>> print(life_stage_info)
    ('Children', '1–3 y')
    """
    
    life_stage_categories = ['Infants', 'Children', 'Males', 'Females', 'Pregnancy', 'Lactation']
    age_ranges = {
        'Infants': ['0–6 mo', '6–12 mo'],
        'Children': ['1–3 y', '4–8 y'],
        'Males': ['9–13 y', '14–18 y', '19–30 y', '31–50 y', '51–70 y', '> 70 y'],
        'Females': ['9–13 y', '14–18 y', '19–30 y', '31–50 y', '51–70 y', '> 70 y'],
        'Pregnancy': ['14–18 y', '19–30 y', '31–50 y'],
        'Lactation': ['14–18 y', '19–30 y', '31–50 y']
    }

    # Prompt user to select their life stage category
    print("Please select your life stage category:")
    for i, category in enumerate(life_stage_categories):
        print(f"{i + 1}. {category}")
    category_choice = int(input("Enter the number of your choice: "))
    selected_category = life_stage_categories[category_choice - 1]

    # Prompt user to select their age range within the life stage category
    print(f"Please select your age range for the '{selected_category}' category:")
    for i, age_range in enumerate(age_ranges[selected_category]):
        print(f"{i + 1}. {age_range}")
    age_choice = int(input("Enter the number of your choice: "))
    selected_age_range = age_ranges[selected_category][age_choice - 1]

    return selected_category, selected_age_range



src/intakeinformer/visualization.py
import matplotlib.pyplot as plt
import pandas as pd
import os

def load_dri_combined():
    """
    Loads the combined DRI DataFrame from a pickle file.
    
    Ensure df_combined is loaded from the pickle file installed alongside all other dependencies and available for use when needed.
    
    Returns
    -------
    pandas.DataFrame
        The DataFrame containing combined DRI data.
    """
    # Construct the relative path to the pickle file
    package_directory = os.path.dirname(__file__)  # Directory of the current file
    relative_path = os.path.join(package_directory, 'data', 'dri_combined.pkl')

    # Load the DataFrame from the pickle file
    df_combined = pd.read_pickle(relative_path)
    return df_combined

df_combined = load_dri_combined()

def plot_calorie_comparison(calorie_data):
    """
    Generates and displays a bar chart for comparing the calorie content of selected food items.

    Serves as an important visualization helper function for IntakeInformer's functionality One - calculating calorie intakes
    
    Parameters
    ----------
    calorie_data : dict
        A dictionary where keys are food item names and values are their corresponding calorie content per 100g. 
        It should also contain a 'total' key for the sum of calories.

    Returns
    -------
    None
        This function displays a bar chart.

    Examples
    --------
    >>> calorie_data = {'BANANA FREEZE DRIED FRUIT': 2000.0, 'GALA APPLES': 28.57, 'ORGANIC VALENCIA ORANGES': 29.22, 'total': 2057.79}
    >>> plot_calorie_comparison(calorie_data)
    # This will display a bar chart comparing the calorie content of the listed food items.
    """
    if 'total' in calorie_data:
        total_calories = calorie_data.pop('total')
    
    food_items = list(calorie_data.keys())
    calories = [calorie_data[item] for item in food_items]

    plt.figure(figsize=(10, 6))
    plt.bar(food_items, calories, color='skyblue')
    plt.xlabel('Food Items')
    plt.ylabel('Calories per 100g')
    plt.title('Comparison of Calorie Content')
    plt.xticks(rotation=45)
    plt.axhline(y=total_calories, color='r', linestyle='-', label=f'Total Calories: {total_calories:.2f}')
    plt.legend()
    plt.show()




def plot_nutrient_comparison(actual_nutrients, df_combined, life_stage_info):
   """
    Generates and displays a bar chart comparing actual nutrient intake against 
    Dietary Reference Intake (DRI) recommendations for a specific life stage.

    Serves as an essential helper function for  IntakeInformer's functionality Two - benchmarking against DRI nutrition allowances.

    Parameters
    ----------
    actual_nutrients : dict
        A dictionary containing the actual nutrient values of the selected food item. 
        Each key-value pair corresponds to a nutrient and its intake value.
    df_combined : pandas.DataFrame
        A DataFrame consolidating DRI values for various nutrients across different life stages.
    life_stage_info : tuple
        A tuple containing the life stage category (e.g., 'Children') and the specific 
        age range (e.g., '1–3 y') for which DRI values are to be compared.

    Returns
    -------
    None
        Displays a bar chart visualization but does not return any value.

    Examples
    --------
    >>> actual_nutrients = {'Carbohydrate(g/d)': 30, 'Proteinb(g/d)': 15, 'Fat(g/d)': 10}
    >>> life_stage_info = ('Children', '1–3 y')
    >>> plot_nutrient_comparison(actual_nutrients, df_combined, life_stage_info)
    [Bar chart is displayed showing the comparison between actual nutrient intake and DRI]

    """
    life_stage_category, age_range = life_stage_info

    # Find the start index for the life stage category
    start_index = df_combined[df_combined['Life StageGroup'] == life_stage_category].index[0]

    # Locate the row for the specific age range within the category
    life_stage_row = df_combined.loc[start_index:].query("`Life StageGroup` == @age_range").iloc[0]

    # Extract DRI values for each nutrient in actual_nutrients
    dri_nutrients = {}
    for nutrient in actual_nutrients.keys():
        # If the nutrient is found in the DRI data and is not NaN, use its value; otherwise, use 0
        dri_value = life_stage_row.get(nutrient)
        dri_nutrients[nutrient] = 0 if dri_value is None or np.isnan(dri_value) else dri_value

    # Filter out nutrients with None values in actual_nutrients
    actual_nutrients = {k: v for k, v in actual_nutrients.items() if v is not None}

    # Define the labels, positions, and width for the bars
    labels = actual_nutrients.keys()
    actual_values = actual_nutrients.values()
    dri_values = [dri_nutrients.get(key, 0) for key in labels]  # Get DRI values or default to 0
    x = range(len(labels))  # the label locations
    width = 0.35  # the width of the bars

    fig, ax = plt.subplots(figsize=(10, 6))
    rects1 = ax.bar(x, actual_values, width, label='Actual')
    rects2 = ax.bar([p + width for p in x], dri_values, width, label='DRI')

    # Add some text for labels, title and custom x-axis tick labels, etc.
    ax.set_ylabel('Nutrient Amount')
    ax.set_title(f'Comparison of Actual Intake vs. DRI Recommendations for {age_range} in {life_stage_category}')
    ax.set_xticks([p + width / 2 for p in x])
    ax.set_xticklabels(labels, rotation=45, ha='right')
    ax.legend()

    def autolabel(rects):
        for rect in rects:
            height = rect.get_height()
            ax.annotate('{}'.format(height),
                        xy=(rect.get_x() + rect.get_width() / 2, height),
                        xytext=(0, 3),  # 3 points vertical offset
                        textcoords="offset points",
                        ha='center', va='bottom')

    autolabel(rects1)
    autolabel(rects2)

    fig.tight_layout()

    plt.show()


src/intakeinformer/intakeinformer.py
import requests
import os
import json
import matplotlib.pyplot as plt

from bs4 import BeautifulSoup
import pandas as pd

FDC_key = os.getenv('POETRY_FDC_API_KEY')

def catbind(a, b):
    concatenated = pd.concat([pd.Series(a.astype("str")),
                              pd.Series(b.astype("str"))])
    return pd.Categorical(concatenated)
'''
internal helper function to fetch data from API and parse it into dataframe for future operation
'''

def _fetch_food_data(query):
    """
    Fetches food data from the USDA FoodData Central API using an internal API key.
    :param query: Food query.
    :return: DataFrame with food data.
    """
    url_search = "https://api.nal.usda.gov/fdc/v1/foods/search"
    params_search = {
        "api_key": FDC_key,
        "pageSize": 10,     # number of results to return, default to 1 for ease of access
        "pageNumber": 2,   # Page number to retrieve, default to 1
        "dataType": [ "Branded"],
        "sortBy": "publishedDate",
        "sortOrder": "asc",
        "query": query
    }
    try:
    # Send a GET request
        response_food_search = requests.get(url_search, params=params_search)

    # If the response was successful, no Exception will be raised
        response_food_search.raise_for_status()
    except response_food_search.exceptions.HTTPError as http_err:
    # Specific HTTP error handling
        print(f'HTTP error occurred: {http_err}')
    except Exception as err:
    # General error handling
        print(f'Other error occurred: {err}')
    else:
    # Success
        data = response_food_search.json()

    if 'foods' in data:
        '''
        pd.json_normalize is used to flatten the nested JSON structure into a DataFrame. 
        '''
        return pd.json_normalize(data, record_path=['foods'])
    else:
        return pd.DataFrame()


def get_user_food_selection(food_df):
    """
    Presents a list of foods to the user and asks them to select one.
    :param food_df: DataFrame with food items.
    :return: The selected food item as a DataFrame row.
    """
    if food_df.empty:
        print("No foods found.")
        return None

    # Display the top results to the user
    for i, row in food_df.iterrows():
        brand_owner = row.get('brandOwner', 'No brand information')
        print(f"{i + 1}: {row['description']} ({brand_owner}, {row.get('servingSize', 'N/A')}g)")

    # Ask the user to make a selection
    try:
        choice = int(input("Enter the number of your choice: ")) - 1
    except ValueError:
        print("Invalid input. Please enter a number.")
        return None

    # Return the selected food item
    if 0 <= choice < len(food_df):
        return food_df.iloc[choice]
    else:
        print("Invalid selection.")
        return None

def calculate_calories_from_selection(selected_items):
    """
    Calculates the calorie content per 100g for user-selected food items and returns a dictionary with individual and total calories.

    Parameters
    ----------
    selected_items : list
        A list of DataFrame rows, each representing a user-selected food item. Each row should contain the food nutrients data.

    Returns
    -------
    dict
        A dictionary where each key is the food item's name, and the corresponding value is its calorie content per 100g. 
        The dictionary also includes a 'total' key representing the sum of calories from all selected items.

    Examples
    --------
    # Assuming selected_banana, selected_apple, and selected_orange are DataFrame rows obtained from user selections
    >>> selected_items = [selected_banana, selected_apple, selected_orange]
    >>> calories_content = calculate_calories_from_selection(selected_items)
    >>> print(calories_content)
    {'Banana': 89.0, 'Apple': 52.0, 'Orange': 47.0, 'total': 188.0}
    """

    calories_dict = {}
    total_calories = 0

    # Nutrient IDs
    ENERGY_ID = 1008
    CARBS_ID = 1005
    FAT_ID = 1004
    PROTEIN_ID = 1003

    for item in selected_items:
        food_name = item['description']
        food_calories = 0  # Initialize calories for this food item

        # Extract the nutrients data
        nutrients_df = pd.json_normalize(item['foodNutrients'])
        nutrients = nutrients_df.set_index('nutrientId').to_dict('index')

        # Get the serving size in grams
        serving_size = item.get('servingSize', 100)  # Default to 100g if not specified

        # Calculate calories per 100g
        if ENERGY_ID in nutrients and 'value' in nutrients[ENERGY_ID]:
            energy_per_serving = nutrients[ENERGY_ID]['value']
            food_calories = (energy_per_serving / serving_size) * 100
        else:
            # Fallback to calculating from protein, fat, and carbohydrates
            for nutrient_id, multiplier in zip([CARBS_ID, FAT_ID, PROTEIN_ID], [4, 9, 4]):
                if nutrient_id in nutrients and 'value' in nutrients[nutrient_id]:
                    nutrient_value_per_100g = (nutrients[nutrient_id]['value'] / serving_size) * 100
                    food_calories += nutrient_value_per_100g * multiplier

        # Add to the total and the dictionary
        total_calories += food_calories
        calories_dict[food_name] = food_calories

    # Add total calories to the dictionary
    calories_dict["total"] = total_calories

    return calories_dict



def plot_calorie_comparison(calorie_data):
    """
    Generates and displays a bar chart for comparing the calorie content of selected food items.

    Parameters
    ----------
    calorie_data : dict
        A dictionary where keys are food item names and values are their corresponding calorie content per 100g. 
        It should also contain a 'total' key for the sum of calories.

    Returns
    -------
    None
        This function displays a bar chart.

    Examples
    --------
    >>> calorie_data = {'BANANA FREEZE DRIED FRUIT': 2000.0, 'GALA APPLES': 28.57, 'ORGANIC VALENCIA ORANGES': 29.22, 'total': 2057.79}
    >>> plot_calorie_comparison(calorie_data)
    # This will display a bar chart comparing the calorie content of the listed food items.
    """
    if 'total' in calorie_data:
        total_calories = calorie_data.pop('total')
    
    food_items = list(calorie_data.keys())
    calories = [calorie_data[item] for item in food_items]

    plt.figure(figsize=(10, 6))
    plt.bar(food_items, calories, color='skyblue')
    plt.xlabel('Food Items')
    plt.ylabel('Calories per 100g')
    plt.title('Comparison of Calorie Content')
    plt.axhline(y=total_calories, color='r', linestyle='-', label=f'Total Calories: {total_calories:.2f}')
    plt.legend()
    plt.show()



def get_calories_for_food_query(*food_queries):
    """
    Handles fetching food data, user selection, and calculating calorie information per 100g for given food queries.

    Parameters
    ----------
    *food_queries : str
        Variable number of string arguments, each representing a food query.

    Returns
    -------
    dict
        A dictionary containing the calorie information for each selected food item per 100g, 
        as well as the total sum of calories.

    Examples
    --------
    >>> get_calories_for_food_query("banana", "apple", "orange")
    {'BANANA FREEZE DRIED FRUIT': 2000.0, 'GALA APPLES': 28.57, 'ORGANIC VALENCIA ORANGES': 29.22, 'total': 2057.79}
    """
    selected_items = []
    for query in food_queries:
        # Fetch food data
        food_df = _fetch_food_data(query)

        # Get user's selection for each query
        print(f"Select an item for '{query}':")
        selected_food = get_user_food_selection(food_df)
        if selected_food is not None:
            selected_items.append(selected_food)
        else:
            print(f"No valid selection made for '{query}'. Skipping...")

    # Calculate calories for the selected items and plot the results
    if selected_items:
        calorie_info = calculate_calories_from_selection(selected_items)
        plot_calorie_comparison(calorie_info)  # Adding visualization here
        return calorie_info
    else:
        return "No calorie information available. No valid selections were made."








src/intakeinformer/data_processing.py
import pandas as pd
import re


def calculate_calories_from_selection(selected_items):
    """
    Calculates the calorie content per 100g for user-selected food items and returns a dictionary with individual and total calories.
    It first tries to use the 'Energy' nutrient value (Energy_id = 1008) and falls back to calculating calories from protein, fat, and carbohydrates if 'Energy' is not located.

    Parameters
    ----------
    selected_items : list
        A list of DataFrame rows, each representing a user-selected food item. Each row should contain the food nutrients data.

    Returns
    -------
    dict
        A dictionary where each key is the food item's name, and the corresponding value is its calorie content per 100g. 
        The dictionary also includes a 'total' key representing the sum of calories from all selected items.

    Examples
    --------
    # Assuming selected_banana, selected_apple, and selected_orange are DataFrame rows obtained from user selections
    >>> selected_items = [selected_banana, selected_apple, selected_orange]
    >>> calories_content = calculate_calories_from_selection(selected_items)
    >>> print(calories_content)
    {'Banana': 89.0, 'Apple': 52.0, 'Orange': 47.0, 'total': 188.0}
    """

    calories_dict = {}
    total_calories = 0

    # Nutrient IDs
    ENERGY_ID = 1008
    CARBS_ID = 1005
    FAT_ID = 1004
    PROTEIN_ID = 1003

    for item in selected_items:
        food_name = item['description']
        food_calories = 0  # Initialize calories for this food item

        # Extract the nutrients data
        nutrients_df = pd.json_normalize(item['foodNutrients'])
        nutrients = nutrients_df.set_index('nutrientId').to_dict('index')

        # Get the serving size in grams
        serving_size = item.get('servingSize', 100)  # Default to 100g if not specified

        # Calculate calories per 100g
        if ENERGY_ID in nutrients and 'value' in nutrients[ENERGY_ID]:
            energy_per_serving = nutrients[ENERGY_ID]['value']
            food_calories = (energy_per_serving / serving_size) * 100
        else:
            # Fallback to calculating from protein, fat, and carbohydrates
            for nutrient_id, multiplier in zip([CARBS_ID, FAT_ID, PROTEIN_ID], [4, 9, 4]):
                if nutrient_id in nutrients and 'value' in nutrients[nutrient_id]:
                    nutrient_value_per_100g = (nutrients[nutrient_id]['value'] / serving_size) * 100
                    food_calories += nutrient_value_per_100g * multiplier

        # Add to the total and the dictionary
        calories_dict[food_name] = round(food_calories, 2)
        total_calories += food_calories

    # Add total calories to the dictionary
    calories_dict["total"] = round(total_calories, 2)

    return calories_dict




def extract_nutrients(food_nutrients):
    """
    Extracts macronutrient information from the foodNutrients data structure, which is the output of get_user_food_selection() function.
    
    Parameters
    ----------
    food_nutrients : list
        A list of dictionaries, each containing nutrient information for a food item.
    
    Returns
    -------
    dict
        A dictionary with the macronutrients and their corresponding values.
    """
    
    # Define the regex patterns for the macronutrients
    patterns = {
        'Carbohydrate(g/d)': re.compile(r'carbohydrate, by difference', re.I),
        'Total Fiber(g/d)': re.compile(r'fiber, total dietary', re.I),
        'Fat(g/d)': re.compile(r'total lipid \(fat\)', re.I),
        'Proteinb(g/d)': re.compile(r'protein', re.I),
        'Calcium (mg/d)': re.compile(r'calcium, ca', re.I),
        'Iron (mg/d)': re.compile(r'iron, fe', re.I),
        'Potassium (mg/d)': re.compile(r'potassium, k', re.I),
        'Sodium (mg/d)': re.compile(r'sodium, na', re.I),
        'Vitamin A(μg/d)a': re.compile(r'vitamin a, (?:IU|μg)', re.I),
        'Vitamin C(mg/d)': re.compile(r'vitamin c, total ascorbic acid', re.I),
        'Vitamin D(μg/d)b,c': re.compile(r'vitamin d(?:3)?', re.I),
        'Vitamin B12(μg/d)': re.compile(r'vitamin b-?12', re.I)
    }
    
    # Initialize a dictionary to hold the macronutrients
    macronutrients = {key: None for key in patterns.keys()}
    
    # Search for the macronutrients in the food_nutrients data
    for nutrient in food_nutrients:
        nutrient_name = nutrient.get('nutrientName', '').lower()
        for key, pattern in patterns.items():
            if pattern.search(nutrient_name):
                # Convert value to float and store in the dictionary
                macronutrients[key] = float(nutrient.get('value', 0))
                break  # Stop searching after finding the nutrient
    
    return macronutrients

